pkg/engine/task/task_pipe.go

```diff
@@ -0,0 +1,427 @@
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/yaml"
+)
+
+var (
+	pipeTaskError = "PipeTaskError"
+
+	pipePodContainerName = "waiter"
+)
+
+type PipeTask struct {
+	Name      string
+	Container string
+	PipeFiles []PipeFile
+}
+
+type PipeFile struct {
+	File string
+	Kind string
+	Key  string
+}
+
+func (pt PipeTask) Run(ctx Context) (bool, error) {
+	// 1. - Render container template -
+	rendered, err := render([]string{pt.Container}, ctx.Templates, ctx.Parameters, ctx.Meta)
+	if err != nil {
+		return false, fatalExecutionError(err, taskRenderingError, ctx.Meta)
+	}
+
+	// 2. - Create core/v1 container object -
+	container, err := unmarshal(rendered[pt.Container])
+	if err != nil {
+		return false, fatalExecutionError(err, resourceUnmarshalError, ctx.Meta)
+	}
+
+	// 3. - Validate the container object -
+	err = validate(container, pt.PipeFiles)
+	if err != nil {
+		return false, fatalExecutionError(err, resourceValidationError, ctx.Meta)
+	}
+
+	// 4. - Create a pod using the container -
+	podName := podName(ctx)
+	podStr, err := pipePod(container, podName)
+	if err != nil {
+		return false, fatalExecutionError(err, pipeTaskError, ctx.Meta)
+	}
+
+	// 5. - Kustomize pod with metadata
+	podObj, err := kustomize(map[string]string{"pipe-pod.yaml": podStr}, ctx.Meta, ctx.Enhancer)
+	if err != nil {
+		return false, fatalExecutionError(err, taskEnhancementError, ctx.Meta)
+	}
+
+	// 6. - Apply pod using the client -
+	podObj, err = apply(podObj, ctx.Client)
+	if err != nil {
+		return false, err
+	}
+
+	// 7. - Wait for the pod to be ready -
+	err = isHealthy(podObj)
+	// once the pod is Ready, it means that its initContainer finished successfully and we can copy
+	// out the generated files. An error during a health check is not treated as task execution error
+	if err != nil {
+		return false, nil
+	}
+
+	// 8. - Copy out the pipe files -
+	log.Printf("PipeTask: %s/%s copying pipe files", ctx.Meta.InstanceNamespace, ctx.Meta.InstanceName)
+	fs := afero.NewMemMapFs()
+	pipePod := podObj[0].(*corev1.Pod)
+
+	err = copyFiles(fs, pt.PipeFiles, pipePod, ctx)
+	if err != nil {
+		return false, err
+	}
+
+	// 9. - Create k8s artifacts (ConfigMap/Secret) from the pipe files -
+	log.Printf("PipeTask: %s/%s creating pipe artifacts", ctx.Meta.InstanceNamespace, ctx.Meta.InstanceName)
+	artStr, err := pipeFiles(fs, pt.PipeFiles, ctx)
+	if err != nil {
+		return false, err
+	}
+
+	// 10. - Kustomize artifacts -
+	artObj, err := kustomize(artStr, ctx.Meta, ctx.Enhancer)
+	if err != nil {
+		return false, fatalExecutionError(err, taskEnhancementError, ctx.Meta)
+	}
+
+	// 11. - Apply artifacts using the client -
+	_, err = apply(artObj, ctx.Client)
+	if err != nil {
+		return false, err
+	}
+
+	// 12. - Delete pipe pod -
+	log.Printf("PipeTask: %s/%s deleting pipe pod", ctx.Meta.InstanceNamespace, ctx.Meta.InstanceName)
+	err = delete(podObj, ctx.Client)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
```

Comment:
so this is basically - create pod, run task.apply and then copy files out and run task.delete - do you think it makes sense to reuse the tasks or rather not?

Reply:
hm. no, I don't think so. the `podObj` is already constructed (no need to read/render/kustomize it first). and the shared part of the logic is confined within the `delete()` call.